### ISSUE TITLE
Lua function "Build" gets stuck when it cannot build units.

### ISSUE THREAD

**Issue by Narg6** [OpenRA/OpenRA#10617]
I was playing around with .Lua AI for custom missions and I found out that the "Build" function can get the AI stuck while training units if they don't have the necessary prerequisites to train that unit.

Example: AI is trying to train 1tnk, 2tnk and arty (Light Tank, Medium Tank and Artillery respectively), then the player destroys the AI's radar, suddenly the AI stops training Tanks, it won't train Light or Medium tanks either, even though they don't need radar to train, they simply get stuck. I managed to workaround it by performing a constant check if the AI has the prerequisites for each unit it will try to build, but I thought this should be mentioned to the devs. Thanks and sorry if that has been posted before.

**Comment by obrakmann** [comment]
That's correct.  The `Build` function creates squads from all actors contained in a single order. It does this by keeping a count of produced actors and comparing that to the number of ordered ones. When a prerequisite to one of those actors is lost, it obviously cannot reach the expected number and the build consequently stops.

You could try working around this by using `CancelProduction()` when the radar gets destroyed, although I haven't tried if that would actually work.

This is unlikely to change (at least in the foreseeable future) since there is no way for the `Build` function to notice that the prerequisite is lost.

**Comment by Narg6** [comment]
To be honest, I would be more interested in seeing the "HackyAI" available for us to use on our Custom Missionss, but I don't know if that would be too hard to implement :-/

**Comment by RReumerman** [comment]
Hi,

I noticed this happening when barracks were destroyed and subsequently rebuilt. "CancelProduction()" is not on the API. Is it new or am I looking at an antiquated version? https://github.com/OpenRA/OpenRA/wiki/Lua-API

How do I use it?

**Comment by obrakmann** [comment]
Yeah, you're right. I thought we had it, but we don't.

**Comment by RReumerman** [comment]
That would be a very useful thing to have for this kind of occasion. 

I ended up writing a bespoke unit training function using the Produce() function but it's likely to break if anything changes in cash or resource management. On the upside, I'm getting pretty close to a mission AI though..
